trace_decoder/src/imp.rs

```diff
@@ -0,0 +1,641 @@
+                        true => receipt.status,
+                        false => true,
+                    };
+
+                let trim_storage = storage_masks.entry(addr).or_default();
+
+                trim_storage.extend(
+                    storage_written
+                        .keys()
+                        .chain(&storage_read)
+                        .map(|it| TrieKey::from_hash(keccak_hash::keccak(it))),
+                );
+
+                if do_writes {
+                    acct.balance = balance.unwrap_or(acct.balance);
+                    acct.nonce = nonce.unwrap_or(acct.nonce);
+                    acct.code_hash = code_usage
+                        .map(|it| match it {
+                            ContractCodeUsage::Read(hash) => {
+                                batch_contract_code.insert(code.get(hash)?);
+                                anyhow::Ok(hash)
+                            }
+                            ContractCodeUsage::Write(bytes) => {
+                                code.insert(bytes.clone());
+                                let hash = keccak_hash::keccak(&bytes);
+                                batch_contract_code.insert(bytes);
+                                Ok(hash)
+                            }
+                        })
+                        .transpose()?
+                        .unwrap_or(acct.code_hash);
+
+                    if !storage_written.is_empty() {
+                        let storage = match born {
+                            true => storage_tries.entry(keccak_hash::keccak(addr)).or_default(),
+                            false => storage_tries
+                                .get_mut(&keccak_hash::keccak(addr))
+                                .context(format!("missing storage trie for address {addr:x}"))?,
+                        };
+
+                        for (k, v) in storage_written {
+                            let slot = TrieKey::from_hash(keccak_hash::keccak(k));
+                            match v.is_zero() {
+                                // this is actually a delete
+                                true => trim_storage.extend(storage.reporting_remove(slot)?),
+                                false => {
+                                    storage.insert(slot, rlp::encode(&v).to_vec())?;
+                                }
+                            }
+                        }
+                        acct.storage_root = storage.root();
+                    }
+
+                    state_trie.insert_by_address(addr, acct)?;
+                }
+
+                if self_destructed {
+                    storage_tries.remove(&keccak_hash::keccak(addr));
+                    state_mask.extend(state_trie.reporting_remove(addr)?)
+                }
+
+                let precompiled_addresses = address!("0000000000000000000000000000000000000001")
+                    ..address!("000000000000000000000000000000000000000a");
```

Comment:
unrelated, can you add a TODO (for myself if you want) to not forget updating it for polygon pos / cdk chains? Thanks

Reply:
#604